### ISSUE TITLE
Refresh not called by `QgsMapCanvas::setLayersPrivate` when removing layers due to broken `layers == oldLayers` comparison

### ISSUE THREAD

**Issue by manisandro** [qgis/QGIS#31250]
When removing layers, `QgsMapCanvas::setLayers` and then `QgsMapCanvas::setLayersPrivate` is called. The `layers == oldLayers` comparison in

    void QgsMapCanvas::setLayersPrivate( const QList<QgsMapLayer *> &layers )
    {
      QList<QgsMapLayer *> oldLayers = mSettings.layers();
      
      // update only if needed
      if ( layers == oldLayers )
        return;

    [...]

however will always return `true`, because `mSettings.layers()` returns `_qgis_listQPointerToRaw( QgsMapSettings::mLayers )` and `_qgis_listQPointerToRaw` filters null pointer entries.
So in other words, the deleted layer will result in a null `QPointer` in `QgsMapSettings::mLayers`, which is then filtered by `_qgis_listQPointerToRaw`, meaning that `oldLayers` will actually only contain the existing layers, which is equal to `layers`.

All this means that the `layersChanged` signal of `QgsMapCanvas` will not get invoked, as well as the call to `refresh`.

CC @wonder-sk as author of https://github.com/qgis/QGIS/commit/5fc10d62077d6d521f5e7c469852d72d70b8905d

**Comment by manisandro** [comment]
Does not affect master anymore.
